src/storage/mod.rs

```diff
@@ -1104,44 +1105,59 @@ impl<E: Engine> Storage<E> {
-                    let cf = Self::rawkv_cf(&cf)?;
-                    // no scan_count for this kind of op.
-
-                    let key_len = key.len();
-                    snapshot.get_cf(cf, &Key::from_encoded(key))
-                        // map storage::engine::Error -> storage::Error
-                        .map_err(Error::from)
-                        .map(|r| {
-                            if let Some(ref value) = r {
-                                let mut stats = Statistics::default();
-                                stats.data.flow_stats.read_keys = 1;
-                                stats.data.flow_stats.read_bytes = key_len + value.len();
-                                thread_ctx.collect_read_flow(ctx.get_region_id(), &stats);
-                                thread_ctx.collect_key_reads(CMD, 1);
-                            }
-                            r
-                        })
-                })
-                .then(move |r| {
-                    _timer.observe_duration();
-                    r
-                })
-        });
+        let _timer = SCHED_HISTOGRAM_VEC
+            .with_label_values(&[CMD])
+            .start_coarse_timer();
+
+        let (tx, future) = oneshot::channel();
+        let readpool = self.read_pool.clone();
+        let region_id = ctx.get_region_id();
+
+        let cb = box move |(_, snapshot)| {
+            if let Ok(fut) = readpool.future_execute(priority, move |ctxd| {
+                future::result(snapshot)
+                    .map_err(txn::Error::from)
+                    .map_err(Error::from)
+                    .and_then(move |snapshot: E::Snap| {
+                        let mut thread_ctx = ctxd.current_thread_context_mut();
+                        let _t_process = thread_ctx.start_processing_read_duration_timer(CMD);
+                        let cf = Self::rawkv_cf(&cf)?;
+                        // no scan_count for this kind of op.
+
+                        let key_len = key.len();
+                        snapshot.get_cf(cf, &Key::from_encoded(key))
+                            // map storage::engine::Error -> storage::Error
+                            .map_err(Error::from)
+                            .map(|r| {
+                                if let Some(ref value) = r {
+                                    let mut stats = Statistics::default();
+                                    stats.data.flow_stats.read_keys = 1;
+                                    stats.data.flow_stats.read_bytes = key_len + value.len();
+                                    thread_ctx.collect_read_flow(region_id, &stats);
+                                    thread_ctx.collect_key_reads(CMD, 1);
+                                }
+                                r
+                            })
+                    })
+                    .and_then(move |res| {
+                        _timer.observe_duration();
+                        let r = tx.send(res);
```

Comment:
em, seem it must be sent successfully. do we need to panic here?

/cc @overvenus

Reply:
Agree, or the command will be blocked for ever.

Reply:
Does it really helpful to add a timer metric here,  The send method of tx should be fast or blocked, other metrics could be used to reflect the timer metric. And we can remove the metric in another PR.

Reply:
> seem it must be sent successfully. do we need to panic here?

By panicking, we may unable to shutdown tikv gracefully.

-------------

> or the command will be blocked for ever.

It's a raw get, I think clients will timeout.

----------------

> Does it really helpful to add a timer metric here, The send method of tx should be fast or blocked, other metrics could be used to reflect the timer metric. And we can remove the metric in another PR.

It does not record the send method, instead it records from L1108 to L1143, means how long it will spend for taking a snapshot and perform a raw get op.